src/operator/numpy/np_where_op-inl.h

```diff
@@ -0,0 +1,266 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * Copyright (c) 2017 by Contributors
+ * \file np_where_op.cc
+ * \brief Function definition of numpy operator where
+ */
+
+#ifndef MXNET_OPERATOR_NUMPY_NP_WHERE_OP_INL_H_
+#define MXNET_OPERATOR_NUMPY_NP_WHERE_OP_INL_H_
+
+#include <mxnet/operator_util.h>
+#include <algorithm>
+#include <string>
+#include <utility>
+#include <vector>
+#include "../../common/utils.h"
+#include "../mxnet_op.h"
+#include "../mshadow_op.h"
+#include "../operator_common.h"
+#include "np_broadcast_reduce_op.h"
+
+namespace mxnet {
+namespace op {
+
+#define NUMPY_WHERE_MAX_DIM 5
+
+using namespace mshadow;
+
+template<int ndim>
+struct numpy_where_kernel {
+  template<typename CType, typename DType>
+  MSHADOW_XINLINE static void Map(index_t base, OpReqType req, const Shape<ndim> &cstride,
+                                  const Shape<ndim> &xstride, const Shape<ndim> &ystride,
+                                  const Shape<ndim> &oshape, CType *datac, DType *datax,
+                                  DType *datay, DType *out) {
+    Shape<ndim> coord = mxnet_op::unravel(base, oshape);
+    auto cidx = static_cast<index_t>(mxnet_op::dot(coord, cstride));
+    auto xidx = static_cast<index_t>(mxnet_op::dot(coord, xstride));
+    auto yidx = static_cast<index_t>(mxnet_op::dot(coord, ystride));
+    KERNEL_ASSIGN(out[base], req, datac[cidx] != CType(0) ? datax[xidx] : datay[yidx]);
+  }
+};
+
+template<int ndim, bool is_left>
+struct numpy_where_backward_kernel {
+  template<typename CType, typename DType>
+  MSHADOW_XINLINE static void Map(index_t base, OpReqType req,
+                                  const Shape<ndim> &cstride, const Shape<ndim> &oshape,
+                                  CType *datac, DType *datao, DType *grad) {
+    Shape<ndim> coord = mxnet_op::unravel(base, oshape);
+    auto cidx = static_cast<index_t>(mxnet_op::dot(coord, cstride));
+    if (is_left) {
+      KERNEL_ASSIGN(grad[base], req, datac[cidx] != CType(0) ? datao[base] : DType(0));
+    } else {
+      KERNEL_ASSIGN(grad[base], req, datac[cidx] == CType(0) ? datao[base] : DType(0));
+    }
+  }
+};
+
+inline bool NumpyWhereOpShape(const nnvm::NodeAttrs& attrs,
```

Comment:
Put shape/type inference functions in .cc if they are not called from other files.